src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -259,7 +259,11 @@ def _get_bound_expression(self, pos, bound):
             # We have the full type information and the bound is known.
             if bound == "lower":
                 return cursor_type.shape[pos].lower.copy()
-            return cursor_type.shape[pos].upper.copy()
+            # If the upper bound is required and is of ArrayType.Extent type
+            # then we'll have to proceed to construct a call to the UBOUND
+            # intrinsic.
+            if not isinstance(cursor_type.shape[pos].upper, ArrayType.Extent):
```

Comment:
Reading the ArrayBounds dataclass docstring "upper" can only return a "DataNode"